### PR TITLE
blindedpath: Minor bug fixes and feature addition.

### PR DESCRIPTION
**Pull request by ziggie1984**
Fixes 2 bugs:
1. We don't need to strictly correlate the invoice expiry with the min_cltv_delta value.
2. Moreover we should respect the minHTLC constraint when creating the blinded route and moreover when adding a buffer to the channel policy when we try to blur the real policy values. 

~~Planning to also add the `SendToBlindedPath` feature when when the sending node is the introduction point.~~

### CODE REVIEW

File: routing/blindedpath/blinded_path.go [hunk 1]
```diff
@@ -105,13 +114,6 @@ type BuildBlindedPathCfg struct {
 func BuildBlindedPaymentPaths(cfg *BuildBlindedPathCfg) (
 	[]*zpay32.BlindedPaymentPath, error) {
 
-	if cfg.MinFinalCLTVExpiryDelta >= cfg.BlocksUntilExpiry {
-		return nil, fmt.Errorf("blinded path CLTV expiry delta (%d) "+
-			"must be greater than the minimum final CLTV expiry "+
-			"delta (%d)", cfg.BlocksUntilExpiry,
-			cfg.MinFinalCLTVExpiryDelta)
-	}
```
Comment (ellemouton) [inline routing/blindedpath/blinded_path.go:113]: can we put this in it's own commit pls 🙏

Comment (ellemouton) [reply routing/blindedpath/blinded_path.go:113]: one thing, one commit :)

File: routing/blindedpath/blinded_path.go [hunk 2]
```diff
@@ -148,7 +150,7 @@ func BuildBlindedPaymentPaths(cfg *BuildBlindedPathCfg) (
 			continue
 		} else if err != nil {
 			log.Errorf("Not using route (%s) as a blinded path: %v",
-				err)
+				route, err)
```
Comment (ellemouton) [inline routing/blindedpath/blinded_path.go:153]: 🙏  thanks

File: routing/blindedpath/blinded_path.go [hunk 3]
```diff
@@ -435,11 +437,27 @@ func collectRelayInfo(cfg *BuildBlindedPathCfg, path *candidatePath) (
 			}
 		}
 
-		policy, err = cfg.AddPolicyBuffer(policy)
+		if policy.MinHTLCMsat > cfg.ValueMsat {
+			return nil, 0, 0, fmt.Errorf("minHTLC of hop policy "+
+				"larger than payment amt: sentAmt(%v), "+
+				"minHTLC(%v)", cfg.ValueMsat,
+				policy.MinHTLCMsat)
```
Comment (ellemouton) [inline routing/blindedpath/blinded_path.go:444]: 👍

**Review by ellemouton — APPROVED** [review]
lgtm! thanks for the quick fixes.

File: routing/blindedpath/blinded_path.go [hunk 4]
```diff
@@ -82,7 +82,16 @@ type BuildBlindedPathCfg struct {
...
 	// BlocksUntilExpiry is the number of blocks that this blinded path
-	// should remain valid for.
+	// should remain valid for. This is a relative number of blocks. This
+	// number in addition with a potential minimum cltv delta for the last
+	// hop and some block padding will be the payment constraint which is
+	// part of the blinded hop info. Every htlc using the provided blinded
+	// hops cannot have a higher cltv delta otherwise it will get rejected
+	// by the forwarding nodes or the final node.
+	//
+	// This number should at least be greater than the invoice expiry time
```
Comment (yyforyongyu) [inline routing/blindedpath/blinded_path.go:92]: Q: can this value be equal to the invoice's final cltv delta?

File: routing/blindedpath/blinded_path.go [hunk 5]
```diff
@@ -435,11 +437,27 @@ func collectRelayInfo(cfg *BuildBlindedPathCfg, path *candidatePath) (
 			}
 		}
 
-		policy, err = cfg.AddPolicyBuffer(policy)
+		if policy.MinHTLCMsat > cfg.ValueMsat {
+			return nil, 0, 0, fmt.Errorf("minHTLC of hop policy "+
```
Comment (yyforyongyu) [inline routing/blindedpath/blinded_path.go:441]: I think this should wrap an `errInvalidBlindedPath`, like how we handle `if minHTLC > maxHTLC {` below.

File: routing/blindedpath/blinded_path.go [hunk 6]
```diff
@@ -435,11 +437,27 @@ func collectRelayInfo(cfg *BuildBlindedPathCfg, path *candidatePath) (
...
 			return nil, 0, 0, err
 		}
 
+		// We only use the new buffered policy if the new minHTLC value
+		// does not violate the sender amount.
+		//
+		// NOTE: We don't check this for maxHTLC, because the payment
+		// amount can always be splitted using MPP.
+		if bufferPolicy.MinHTLCMsat <= cfg.ValueMsat {
+			policy = bufferPolicy
```
Comment (yyforyongyu) [inline routing/blindedpath/blinded_path.go:458]: IIUC we should not choose this hop if the buffered policy cannot meet our requirement? I think we can do a single check on this `if bufferPolicy.MinHTLCMsat > cfg.ValueMsat {` and return a wrapped `errInvalidBlindedPath` here if `bufferPolicy.MinHTLCMsat` is at least `policy.MinHTLCMsat`

File: routing/blindedpath/blinded_path.go (hunk 4 above)
Comment (ziggie1984) [reply routing/blindedpath/blinded_path.go:92]: Yes this value is not related to the final_cltv value, the final_cltv_value is added on top of this expiry when calculating the payment constraint for each blinded hop.

File: routing/blindedpath/blinded_path.go (hunk 6 above)
Comment (yyforyongyu) [reply routing/blindedpath/blinded_path.go:458]: yeah no strong opinion here (since i don't know it well😂), more like a question so feel free to ignore it.

**Review by yyforyongyu — APPROVED** [review]
LGTM🙏
